### ISSUE TITLE
There seems to be just one thread running.

### ISSUE THREAD

**Issue by pawarvaibhav** [prashant-ramcharan/courgette-jvm#53]
In your example project, I observed that there is always just 1 thread running. 
I added this line of code in your step definition - 

 @When("I navigate to Stack Overflow question page (\\d+)")
    public void navigateToStackOverflowQuestionPage(Integer page) {
        **System.out.println("THREAD NUMBER " + Thread.currentThread().getId());**
        driver.navigate().to("https://stackoverflow.com/questions?page=" + page);
    }

In the output, I always see THREAD NUMBER: 1 for all scenarios. I thought if the project uses multiple threads for execution, if we set threads to 10.

Am I missing anything?

P. S.  I need to access the different thread IDs for some processing.

**Comment by prashant-ramcharan** [comment]
The threads are forked by Gradle. 

When you run the test, it starts a single thread, that thread then start multiple threads, so you will always get the initial thread ID. You will not have access to the threads started by Courgette - this is by design.

How do you know its just ran 1 thread? By just printing out Thread.currentThread().getId()?

Do you no see multiple browsers opening?

http://recordit.co/9Oq6PmlgMg

**Comment by pawarvaibhav** [comment]
Yeah. I did see the multiple browsers. I now understand what is going on. 

The problem I am trying to solve is:
Consider following Scenario
    Scenario: Verify 200 success response code.
        When I send a GET request to "http://this/is/an/API"
        Then the response status should be "200"

In When above, I create a 'storage' object and save the API response.
In Then step, I access the 'storage' object created above and read the response and perform validations.

Is this possible with your framework? I am assuming with your current setup, I cannot create 'storage' objects and tie them to individual threads.

**Comment by prashant-ramcharan** [comment]
Short answer: Yes.

Long answer: You thinking about this the wrong way. You don't need to add extra logic in your tests for it to work with this library. 

If your test passes normally with Cucumber, it will pass with Courgette. Courgette simply takes your existing Cucumber scenario / feature (_in it's entirety_) and runs it in an independent thread. 

Courgette does not run step definitions in parallel; it runs an entire **scenario / feature** in parallel. 

In the below scenario, **all steps** will run in the **same** thread. Test state (i.e. variables / storage classes) is lost when the thread ends so theres no need to cross reference data between threads.

**This does not mean**: 1 thread will run the **When** step and another will run the **Then** step.

````java
private int responseCode;

@When("I send a GET request")
public void send() {
  responseCode = service.call("http://some-api/");
}

@Then("I verify a success response code")
public void verify() {
  assertThat(responseCode).equalTo(200);
}
````
Not sure why you would need to share data between **threads**? You should be sharing data between steps definitions. 

**In summary**: Write your test for a single purpose (_i.e. a single Cucumber test_) - parallelism comes out the box using Courgette. 

Don't write your tests to work with Courgette - write your tests to work with Cucumber. If it works with Cucumber then you will be able to run it in parallel using Courgette.

**Comment by pawarvaibhav** [comment]
Hey Prashant, 

One last question on this topic. I am afraid I am asking you a java question and not a courgette question. What if I need to save step definitions in separate classes and still share data? In your example above, 

`@When("I send a GET request")`
and 
`@Then("I verify a success response code")`

are in the same class.

What if I need them in separate classes? How can I share 
`int responseCode;`
in that case?

We organize step definitions in separate java files, since we have like a generic set of steps and a project specific set of steps.

**Comment by prashant-ramcharan** [comment]
Hi @pawarvaibhav 

Ideally, you should have a SessionContext which you can add to and refer back to from any step --- a static HashMap would suffice.

Although for this example / explanation, lets just use a static variable. 

The variable be available to every class / step def and will only exist in its current thread. There will not be any conflicts with other threads as each thread will hold its own instance of Session.

Then, in your step, simply get / set the value.
````java
Session.responseCode = 200;
````

````java
public abstract class Session {
 
public static int responseCode;

 public static void clear() {
    responseCode = null
   }
}
````

You should reset the Session before or after each cucumber scenario.

````java
@Before
public void beforeScenario() {
   Session.clear();
}
